crates/lsp/src/lsp.rs

```diff
@@ -631,7 +631,8 @@ impl LanguageServer {
                                     "filterText".to_string(),
                                     "labelDetails".to_string(),
                                     "tags".to_string(),
-                                    "textEdit".to_string(),
+                                    // NB: Do not have this resolved, otherwise Zed becomes slow to complete things
+                                    // "textEdit".to_string(),
```

Comment:
Hi, @SomeoneToIgnore - Sorry to jump in here randomly, but I was just going through this PR and was just curious about the comment: `// NB: Do not have this resolved, otherwise Zed becomes slow to complete things`. Why would doing `textEdit.to_string()` here would be a cause (/major cause? - if it is) of slowness?

Reply:
A rhetorical question: why did you have to review this PR, instead of just commenting?

------------------------------

Due to LSP, we have a choice to either receive a lot of JSON text in response for `textDocument/completion` (non-resolved state) or omit certain fields out and send another, `completionItem/resolve`, request to get those fields later.

In the former case, the server might delay sending those completions back due to doing extra work and computing everything. 
Also, the client has to accept all this and potentially parse it on the spot, so more work too.
The issue here is the abuncance of markdown docs and whatever that could occupy up to 40% of the completion response, for nothing, as most of the items will not even be shown in the completion menu (only first ~dozen will, but ALL of the items have to be sent fully if not intended to be resolved).
But after the list is received, all the client has to do is to apply items' properties locally, on selection, when completion, and that's in.

Resolving as much as possible, might trim that 40% off, but forces us to do another request per completion item to get more data back.

Moreover, completions have `text_edit` and `additional_text_edits` and all of them has to be applied after selecting one.
Currently, Zed is acrhitectured so that `text_edit` is being received and parsed into an internal Zed change data
https://github.com/zed-industries/zed/blob/39468de8c6a4d90855cb1c7138d315b6248dabf3/crates/project/src/lsp_command.rs#L1805-L1817
or edit's label is taken instead for the same purpose.

Later, this `text_edit` is instantly applied on accepting the completion and the `additional_text_edits` are following up asynchronously, after being resolved (potentially, but this the case for r-a, for example).
That resolution is happening quite eager inside the `CompletionsMenu`, for every item visible and/or selected, but this seems to be quite slow nonetheless, as it turned out.

So, when we explicitly tell the LSP server to resolve the `textEdit` later, we start to get the edit-less completions that are slowly resolved in the background despite the fact we start to resolve it quite eagerly - so slow that sometimes it takes seconds.

And now I had a choice: to debug and alter a somewhat complicated resolve logic that _seems_ to be doing things correctly (with a hight chance to break something or to find out that everything's right and resolve is indeed slow - also note that I'll need to check that withmultiple various language servers as they are all behaving differently); or to fall back to the original status quo and accept that the server sends us a bit more data a bit slower than we might want, but at least it seems to work for everyone and Zed does not need to query anything vital for completions to apply.

Reply:
> A rhetorical question: why did you have to review this PR, instead of just commenting?

It was a miss from my end, apologies for this. I forgot to put it as a comment instead. Hope it's not a big deal! :) Thanks for pointing it out.

---

Thanks for the thorough response as well and taking the time to explain what could be going on here! Appreciate it.